File: rough_work/assignment_inception_python/mill_observer.py
from game_token import GameToken


class MillObserver:
    "An observer that checks for the formation of mills."

    def __init__(self):
        self.tokens: list[GameToken] = []

    def notify_push(self, token: GameToken) -> None:
        "Notify the observer of a push to the tokens it is watching."

        self.tokens.append(token)

    def notify_pop(self, token: GameToken):
        "Notify an observer of a pop from the tokens it is watching."

        if token in self.tokens:
            self.tokens.remove(token)

    def has_mill(self, colour: str) -> bool:
        "Check if a mill has been formed."

        token_counter = 0

        for token in self.tokens:
            if token.get_colour() == colour:
                token_counter += 1

        if token_counter == 3:
            return True

        return False

    def get_tokens(self) -> list[GameToken]:
        "Get the tokens that being observed by the MillObserver."

        return self.tokens

File: src/gameengine/MillObserver.java
package gameengine;

import java.util.ArrayList;
import tokens.Token;
import tokens.TokenType;

/** An Observer that watches for the formation of Mills. */
public class MillObserver {
  private final ArrayList<Token> tokens = new ArrayList<>();

  /**
   * Check whether the MillObserver is currently observing a Mill of a particular TokenType.
   *
   * @return True if a mill exists, false if not.
   */
  public boolean hasMill() {
    if (this.tokens.size() != 3) {
      return false;
    }

    TokenType tokenType = tokens.get(0).getTokenType();
    int tokensOfGivenType = 0;

    for (Token token : tokens) {
      if (token.getTokenType() == tokenType) {
        tokensOfGivenType += 1;
      }
    }
    
    return (tokensOfGivenType == 3);
  }

  /**
   * Update the MillObserver's Token list by adding a Token.
   *
   * @param token The Token being added.
   */
  public void updateTokenAddition(Token token) {
    this.tokens.add(token);
  }

  /**
   * Update the MillObserver's Token list by removing a Token.
   *
   * @param token The Token being removed.
   */
  public void updateTokenRemoval(Token token) {
    this.tokens.remove(token);
  }
}
